src/meepgeom.cpp

```diff
@@ -1757,44 +1777,80 @@ void fragment_stats::update_stats_from_material(material_type mat, size_t pixels
   }
 }
 
+void fragment_stats::compute_overlaps_from_tree(geom_box_tree t, std::vector<double> &overlaps) {
+  if (!t || !geom_boxes_intersect(&t->b, &box))
+    return;
+
+  if (t->nobjects > 0) {
+    geom_box intersection;
+    geom_box_intersection(&intersection, &t->b, &box);
+
+    for (int i = 0; i < t->nobjects; ++i) {
+      if (geom_boxes_intersect(&intersection, &t->objects[i].box)) {
+        geom_box shifted_box = {vector3_minus(intersection.low, t->objects[i].shiftby),
+                                vector3_minus(intersection.high, t->objects[i].shiftby)};
+        double overlap = box_overlap_with_object(shifted_box, *t->objects[i].o, tol, maxeval);
+        if (overlap > 0.0) {
+          for (int j = 0; j < geom.num_items; ++j) {
+            if (t->objects[i].o == geom.items + j) {
+              overlaps[j] += overlap;
+              break;
+            }
+          }
+        }
+      }
+    }
+  }
+
+  compute_overlaps_from_tree(t->t1, overlaps);
+  compute_overlaps_from_tree(t->t2, overlaps);
+}
+
 void fragment_stats::compute_stats() {
 
   if (geom.num_items == 0) {
     // If there is no geometry, count the default material for the whole fragment
     update_stats_from_material((material_type)default_material, num_pixels_in_box);
   }
+  else {
+    std::vector<double> overlaps(geom.num_items, 0.0);
+    compute_overlaps_from_tree(geom_tree, overlaps);
 
-  for (int i = 0; i < geom.num_items; ++i) {
-    geometric_object *go = &geom.items[i];
-    double overlap = box_overlap_with_object(box, *go, tol, maxeval);
-
+    size_t total_obj_pixels = 0;
     bool anisotropic_pixels_already_added = false;
-    if (eps_averaging) {
-      // If the object doesn't overlap the entire box, that implies that
-      // an object interface intercepts the box, which means we treat
-      // the entire box as anisotropic. This method could give some false
-      // positives if there is another object with the same material behind
-      // the current object, but in practice it is probably reasonable to
-      // assume that there is a material interface somwhere in the box so
-      // we won't worry about fancier edge-detection methods for now.
-      if (overlap != 1.0) {
-        anisotropic_pixels_already_added = true;
-        num_anisotropic_eps_pixels += num_pixels_in_box;
-        if (mu_not_1(go->material)) {
-          num_anisotropic_mu_pixels += num_pixels_in_box;
+
+    for (int i = 0; i < geom.num_items; ++i) {
+      geometric_object *go = &geom.items[i];
+      double overlap = overlaps[i];
```

Comment:
Would be good to check if this roughly matches the overlaps computed for each object with the old code.